torch_geometric/nn/pool/mem_pool.py

```diff
@@ -0,0 +1,100 @@
+from typing import Optional, Tuple
+
+import torch
+from torch import Tensor
+from torch.nn import Parameter, KLDivLoss, Conv2d, Linear
+
+from torch_geometric.utils import to_dense_batch
+
+
+class MemPooling(torch.nn.Module):
+    r"""Memory based pooling layer from `"Memory-Based Graph Networs"
+    <https://arxiv.org/abs/2002.09518>`_ paper, which learns a coarsened graph
+    representation based on soft cluster assignments
+
+    .. math::
+        S_{i,j}^{(h)} &= \frac{
+        (1+{\| \mathbf{x}_i-\mathbf{k}^{(h)}_j \|}^2 / \tau)^{
+        -\frac{1+\tau}{2}}}{
+        \sum_{k=1}^K (1 + {\| \mathbf{x}_i-\mathbf{k}^{(h)}_k \|}^2 / \tau)^{
+        -\frac{1+\tau}{2}}}
+
+        \mathbf{S} &= \textrm{softmax}(\textrm{Conv2d}
+        (\Vert_{h=1}^H \mathbf{S}^{(h)})) \in \mathbb{R}^{N \times K}
+
+        \mathbf{X}^{\prime} &= \mathbf{S}^{\top} \mathbf{X} \mathbf{W} \in
+        \mathbb{R}^{K \times F^{\prime}}
+
+    Where :math:`H` denotes the number of heads, and :math:`K` denotes the
+    number of clusters.
+
+    Args:
+        in_channels (int): Size of each input sample :math:`F`.
+        out_channels (int): Size of each output sample :math:`F^{\prime}`.
+        heads (int): The number of heads :math:`H`.
+        num_clusters (int): number of clusters :math:`K` per head.
+        tau (int, optional): The temperature :math:`\tau`. (default: :obj:`1.`)
+    """
+    def __init__(self, in_channels: int, out_channels: int, heads: int,
+                 num_clusters: int, tau: float = 1.):
+        super(MemPooling, self).__init__()
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.heads = heads
+        self.num_clusters = num_clusters
+        self.tau = tau
+
+        self.k = Parameter(torch.Tensor(heads, num_clusters, in_channels))
+        self.conv = Conv2d(heads, 1, kernel_size=1, padding=0, bias=False)
+        self.lin = Linear(in_channels, out_channels, bias=False)
+
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        torch.nn.init.uniform_(self.k.data, -1., 1.)
+        self.conv.reset_parameters()
+        self.lin.reset_parameters()
+
+    @staticmethod
+    def kl_loss(S: Tensor) -> Tensor:
+        r"""The additional KL divergence-based loss
+
+        .. math::
+            P_{i,j} &= \frac{S_{i,j}^2 / \sum_{n=1}^N S_{n,j}}{\sum_{k=1}^K
+            S_{i,k}^2 / \sum_{n=1}^N S_{n,k}}
+
+            \mathcal{L}_{\textrm{KL}} &= \textrm{KLDiv}(\mathbf{P} \Vert
+            \mathbf{S})
+        """
+        S_2 = S**2
+        P = S_2 / S.sum(dim=1, keepdim=True)
+        P = P / (S_2.sum(dim=2, keepdim=True) / S.sum(dim=1, keepdim=True))
```

Comment:
@rusty1s . shouldn't this line be
```
denom=P.sum(2,keepdim=True)
P=P/denom
```
![image](https://user-images.githubusercontent.com/13963626/111269427-a3e4fc80-8654-11eb-9731-a2a8f74b80a4.png)
In the denominator the inner summation is inside the outer summation.  The equation is probably confusing as they didn't add bracket after the outer summation.